janusgraph-hadoop-parent/janusgraph-hadoop-core/src/main/java/org/janusgraph/hadoop/formats/cassandra/CqlBridgeRecordReader.java

```diff
@@ -0,0 +1,377 @@
+            newKV.addEntries(entries);
+            currentKV = newKV;
+            return true;
+        }
+    }
+
+    /**
+     * Return native version protocol of the cluster connection
+     *
+     * @return serialization protocol version.
+     */
+    public int getNativeProtocolVersion() {
+        return nativeProtocolVersion;
+    }
+
+    /**
+     * A non-static nested class that represents an iterator for distinct keys based on the
+     * row iterator from DataStax driver. In the usual case, more than one row will be associated
+     * with a single key in JanusGraph's use of Cassandra.
+     */
+    private class DistinctKeyIterator implements Iterator<Map<StaticArrayBuffer, Map<StaticBuffer, StaticBuffer>>> {
+        private final Iterator<Row> rowIterator;
+        long totalRead;
+        Row previousRow = null;
+        DistinctKeyIterator() {
+            AbstractType type = partitioner.getTokenValidator();
+            Object startToken = type.compose(type.fromString(split.getStartToken()));
+            Object endToken = type.compose(type.fromString(split.getEndToken()));
+            SimpleStatement statement = new SimpleStatement(cqlQuery, startToken, endToken);
+            rowIterator = session.execute(statement).iterator();
+            for (ColumnMetadata meta : cluster.getMetadata().getKeyspace(quote(keyspace)).getTable(quote(cfName)).getPartitionKey()) {
+                partitionBoundColumns.put(meta.getName(), Boolean.TRUE);
+            }
+        }
+
+        @Override
+        public boolean hasNext() {
+            return rowIterator.hasNext();
+        }
+
+        /**
+         * <p>
+         *     Implements the <i>business logic</i> of the outer class. Refer to {@linkplain CqlBridgeRecordReader}.
+         * Relies on the {@linkplain Iterator} of {@linkplain Row} to get a map of rows that correspond
+         * to the same key.
+         * </p>
+         * <p>
+         *     Note: This is not a general purpose iterator. There is no provision of {@linkplain java.util.ConcurrentModificationException}
+         *     while iterating using this iterator.
+         * </p>
+         * @return the next element in the iteration of distinct keys, returns <code>null</code> to indicate
+         * end of iteration
+         */
+        @Override
+        public Map<StaticArrayBuffer, Map<StaticBuffer, StaticBuffer>> next() {
+            if (! rowIterator.hasNext()) {
+                return null; // null means no more data
+            }
+            Map<StaticArrayBuffer, Map<StaticBuffer, StaticBuffer>> kcvs = new HashMap<>(); // key -> (column1 -> value)
+            Row row;
+            if (previousRow == null) {
+                row = rowIterator.next(); // just the first time, should succeed
+            } else {
+                row = previousRow;
+            }
+            StaticArrayBuffer key = StaticArrayBuffer.of(row.getBytesUnsafe("key"));
```

Comment:
Do you believe they have any reason to exist outside this method? Or is this a JG code style issue?